Add support for SLO

pkarman opened the pull request:
The newest `ruby-saml` library now supports single log out request/response. This PR adds corresponding support for that feature to IDPs.

jphenow commented on lib/saml_idp/request.rb:
```diff
@@ -31,48 +31,129 @@ def initialize(raw_xml = "")
+      if authn_request?
+        authn_request["ID"]
+      elsif logout_request?
+        logout_request["ID"]
```
Because these requests are somewhat similar, thoughts on having a single `request` location where that returns the proper xpath? They seem mutually exclusive as well. Might not need to expose the new boolean methods in that case?

pkarman replied:
Our app uses the boolean methods to determine the flavor of the request. It seems reasonable to me that there be some public way to determine that flavor, since this class is doing double-duty.

I did add a `request` method at your suggestion, in 6ac5ea86e24c2f96b5e60ef25323bce2a3d8161b

jphenow replied:
Yea that crossed my mind as I hit enter on that comment, that seems reasonable. Thanks for the add

jphenow commented on lib/saml_idp/request.rb:
```diff
@@ -31,48 +31,129 @@ def initialize(raw_xml = "")
+      end
+    end
+
+    def logger(msg)
```
logger implies we're returning some sort of log object. Maybe rename to `log` or return a logging object. The non-Rails one could delegate the function down to puts still, if you prefer to return an object

pkarman replied:
good point. method renamed in 6ac5ea86e24c2f96b5e60ef25323bce2a3d8161b

jphenow commented on lib/saml_idp/request.rb:
```diff
@@ -31,48 +31,129 @@ def initialize(raw_xml = "")
-      service_provider? &&
-        valid_signature? &&
-        acs_url.present?
+      # TODO: This should validate against the schema.
```
I agree, but I don't know that an inline comment will get it done, an issue would possibly be a better route.

jphenow commented on lib/saml_idp/request.rb:
```diff
@@ -31,48 +31,129 @@ def initialize(raw_xml = "")
     def issuer
-      @content ||= xpath("//saml:Issuer", saml: assertion).first.try(:content)
-      @content if @content.present?
+      @_issuer ||= xpath("//saml:Issuer", saml: assertion).first.try(:content)
```
I believe there was a reason for the `@content if @content.present?` maybe append a `presence` to the call so we have no opportunity to output a blank string or something.

pkarman replied:
I added back the `.present?` and a test to verify it.

jphenow commented on spec/lib/saml_idp/controller_spec.rb:
```diff
@@ -30,18 +30,28 @@ def params
       response = OneLogin::RubySaml::Response.new(saml_response)
       response.name_id.should == "foo@example.com"
-      response.issuer.should == "http://example.com"
+      response.issuers.first.should == "http://example.com"
```
Do we not still have access to `issuer`?

pkarman replied:
apparently not. I searched awhile to see where ruby-saml dropped that method, but couldn't find it in the changelog.

jphenow replied:
Ah this is from the external OneLogin lib. That's out of our control then. As long as that's not an API change on our part, we're good.

jphenow commented on spec/xml_security_spec.rb:
```diff
@@ -116,7 +116,9 @@ module SamlIdp
         Timecop.freeze Time.parse('2012-11-28 17:55:00 UTC') do
-          response.validate!.should be_truthy
+          response.soft = false
+          skip "starfield_response fixture is not valid. missing SubjectConfirmationData element."
```
can we get this operating again?

jphenow commented:
Thanks for the PR. Just left a few comments. Looking pretty good otherwise.

Are there changes to the code that would break the library for people who were to just update the gem and not their configuration or anything?

pkarman commented:
Thanks for the good review @jphenow -- I think I've addressed all your comments.

I am not aware of any backwards incompatible changes. I'm basing that on the minimal churn in the `spec/` space. It appears to be mostly additive, with a few small changes to match the changes in ruby-saml (the latter was a big version jump).

jphenow commented:
Awesome. Looks good to me. Is the hope to have #43 merged along with this then do a single release? 

cc @anfleene @natejgreene so you guys know I'll be releasing some updates here

pkarman commented:
Yes, hopefully #43 could also be merged and then a single release.

anfleene commented:
@jphenow Awesome sounds good.

smsohan commented:
@pkarman any interest / plans to add the HTTP Redirect binding for SLO? I'd like to give it a shot.

pkarman commented:
@smsohan I do not maintain this repository, so can't comment on the addition of new code here.

Because the maintainers here have been unresponsive, we forked and maintain our own copy at https://github.com/18F/saml_idp
